### PR TITLE
Feat: add DB2 test dispatch + per-db guards

### PR DESCRIPTION
**Pull request by paulteehan**
## Summary
- Add `db2` branch in `DataSourceTestHelper.create()` for test dispatch
- Add `db2` to `cursor.rowcount = -1` list in `test_query_result_iterator`
- Add `tz_unaware_data_sources` branch in `test_full_create_insert_drop_ast` for DB2's lack of TIMESTAMP WITH TIME ZONE

## Context
These are test-only changes needed to run the canonical soda-tests suite against DB2 LUW. The adapter itself lives in soda-extensions (sodadata/soda-extensions#346).

## Test plan
- [ ] Verify no regressions on existing adapters (postgres, snowflake, etc.)
- [ ] CI run with `TEST_DATASOURCE=db2` once soda-extensions adapter is installed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: soda-tests/tests/integration/test_sql_ast_inserts.py [hunk 1]
```diff
@@ -231,10 +231,22 @@ def col_type(name: str) -> str:
...
+                2021, 1, 1, 10, 0, 0, tzinfo=datetime.timezone.utc
+            )
+            assert interpret_datetime_as_utc(result.rows[1][5]) == datetime.datetime(
+                2021, 1, 2, 10, 0, 0, tzinfo=datetime.timezone.utc
+            )
+        else:
+            assert interpret_datetime_as_utc(result.rows[0][5]) == datetime.datetime(
+                2021, 1, 1, 10, 0, 0, tzinfo=datetime.timezone.utc
+            )
+            assert interpret_datetime_as_utc(result.rows[1][5]) == tz.localize(datetime.datetime(2021, 1, 2, 10, 0, 0))
```
Comment (Niels-b) [inline soda-tests/tests/integration/test_sql_ast_inserts.py]: Unsure why we need this 🤔 . There are other datasources that do something similar, no? What exactly makes this different such that we need an edge case?

Comment (paulteehan) [reply soda-tests/tests/integration/test_sql_ast_inserts.py]: Thanks, I missed this, this should have been handled in the dialect, fixed.

**Review by Niels-b — APPROVED** [review]
